test/lit/passes/remove-unused-brs_all-features.wast

```diff
@@ -116,110 +118,38 @@
   )
  )
 
- ;; CHECK:      (func $br_on-to-br (type $ref|func|_=>_none) (param $func (ref func))
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 0)
- ;; CHECK-NEXT:  )
+ ;; CHECK:      (func $br_on_null (type $none_=>_none)
  ;; CHECK-NEXT:  (block $null
  ;; CHECK-NEXT:   (drop
- ;; CHECK-NEXT:    (ref.func $br_on-to-br)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:   (call $log
- ;; CHECK-NEXT:    (i32.const 1)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 2)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (drop
- ;; CHECK-NEXT:   (block $func (result (ref $ref|func|_=>_none))
- ;; CHECK-NEXT:    (drop
- ;; CHECK-NEXT:     (br $func
- ;; CHECK-NEXT:      (ref.func $br_on-to-br)
- ;; CHECK-NEXT:     )
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (call $log
- ;; CHECK-NEXT:     (i32.const 3)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (ref.func $br_on-to-br)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 4)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (drop
- ;; CHECK-NEXT:   (block $i31 (result (ref i31))
- ;; CHECK-NEXT:    (drop
- ;; CHECK-NEXT:     (br $i31
- ;; CHECK-NEXT:      (i31.new
- ;; CHECK-NEXT:       (i32.const 42)
- ;; CHECK-NEXT:      )
- ;; CHECK-NEXT:     )
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (call $log
- ;; CHECK-NEXT:     (i32.const 5)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (i31.new
- ;; CHECK-NEXT:     (i32.const 1337)
+ ;; CHECK-NEXT:    (br_on_null $null
+ ;; CHECK-NEXT:     (ref.null nofunc)
  ;; CHECK-NEXT:    )
  ;; CHECK-NEXT:   )
  ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (call $log
- ;; CHECK-NEXT:   (i32.const 6)
- ;; CHECK-NEXT:  )
- ;; CHECK-NEXT:  (drop
- ;; CHECK-NEXT:   (block $non-null (result (ref $ref|func|_=>_none))
- ;; CHECK-NEXT:    (br $non-null
- ;; CHECK-NEXT:     (ref.func $br_on-to-br)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (call $log
- ;; CHECK-NEXT:     (i32.const 7)
- ;; CHECK-NEXT:    )
- ;; CHECK-NEXT:    (ref.func $br_on-to-br)
- ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:  )
  ;; CHECK-NEXT: )
- (func $br_on-to-br (param $func (ref func))
-  (call $log (i32.const 0))
+ (func $br_on_null
   (block $null
-   ;; a non-null reference is not null, and the br is never taken
+   ;; a null reference to bottom is definitely null, and the br is always taken
+   ;; TODO: Optimize this.
```

Comment:
Please add a test for the reverse case, which is optimized and which we had a test for before, where the input is non-null so the br is never taken.

Reply:
Updated. It seems we don't yet optimize in the known-null case for either br_on_null or br_on_non_null.